### PR TITLE
Plugin procedural assemblies

### PR DESCRIPTION
**Pull request by est77**
sphereflake
Sphereflake fractal sample assembly plugin

### CODE REVIEW

File: src/appleseed/renderer/modeling/scene/pluginassembly.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        auto_release_ptr<IAssemblyFactory> assembly_factory(create_fn());
+
+        // Create the plugin assembly.
+        plugin_assembly = assembly_factory->create(
+            get_name(),
+            get_parameters().child("parameters"));
+    }
+    catch (const ExceptionCannotLoadSharedLib& e)
+    {
+        RENDERER_LOG_ERROR("cannot open plugin assembly: %s", e.what());
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/pluginassembly.cpp]: Add assembly path to error message.

File: src/appleseed/renderer/modeling/scene/pluginassembly.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        return false;
+    }
+    catch (const ExceptionPluginInitializationFailed&)
+    {
+        RENDERER_LOG_ERROR("initialization of assembly plugin %s failed", plugin_path.c_str());
+        return false;
+    }
+    catch (const ExceptionSharedLibCannotGetSymbol& e)
+    {
+        RENDERER_LOG_ERROR("cannot load symbol %s from assembly plugin", e.what());
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/pluginassembly.cpp]: Add assembly path to error message.

File: src/appleseed/renderer/modeling/scene/pluginassembly.cpp [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        return false;
+    }
+
+    // Expand contents of the plugin assembly.
+    const bool result = static_cast<PluginAssembly*>(plugin_assembly.get())->expand_contents(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/pluginassembly.cpp]: `result` -> `success`?

File: sandbox/samples/cpp/sphereflake/sphereflake.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+// appleseed.foundation headers.
+#include "foundation/math/vector.h"
+#include "foundation/utility/string.h"
+
+// Standard headers.
+#include <cstring>
+#include <memory>
+
+// Sphereflake headers.
+#include"sphereflakedata.h"
```
Comment (dictoon) [inline sandbox/samples/cpp/sphereflake/sphereflake.cpp]: Add space after `#include`.

File: sandbox/samples/cpp/sphereflake/sphereflake.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+class SphereFlakeAssembly
+  : public asr::PluginAssembly
+{
+  public:
+    SphereFlakeAssembly(const char* name, const asr::ParamArray params)
```
Comment (dictoon) [inline sandbox/samples/cpp/sphereflake/sphereflake.cpp]: Pass `ParamArray` by const ref.

File: sandbox/samples/cpp/sphereflake/sphereflake.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    bool expand_contents(
+        const asr::Project&     project,
+        const asr::Assembly*    parent,
+        asf::IAbortSwitch*      abort_switch = 0) override
+    {
+        build_sphere_master_object();
+
+        const double *p = spheres;
```
Comment (dictoon) [inline sandbox/samples/cpp/sphereflake/sphereflake.cpp]: Move `*` to type's side.

File: sandbox/samples/cpp/sphereflake/sphereflake.cpp [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        for (size_t i = 0; i < num_spheres; ++i)
+        {
+            const asf::Vector3d center(p[0],p[1],p[2]);
```
Comment (dictoon) [inline sandbox/samples/cpp/sphereflake/sphereflake.cpp]: Add space after commas.

File: sandbox/samples/cpp/sphereflake/sphereflake.cpp (hunk 4 above)
Comment (dictoon) [reply sandbox/samples/cpp/sphereflake/sphereflake.cpp]: Typically the sphere snowflake is generated procedurally no? :)

File: sandbox/samples/cpp/sphereflake/sphereflake.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    bool expand_contents(
+        const asr::Project&     project,
+        const asr::Assembly*    parent,
+        asf::IAbortSwitch*      abort_switch = 0) override
+    {
+        build_sphere_master_object();
+
+        const double *p = spheres;
+        for (size_t i = 0; i < num_spheres; ++i)
```
Comment (dictoon) [inline sandbox/samples/cpp/sphereflake/sphereflake.cpp]: `NumSpheres` if it's a constant.

File: sandbox/samples/cpp/sphereflake/sphereflakedata.h [hunk 9]
```diff
@@ -0,0 +1,7417 @@
+
```
Comment (dictoon) [inline sandbox/samples/cpp/sphereflake/sphereflakedata.h]: Remove one blank line.

File: sandbox/samples/cpp/sphereflake/sphereflake.cpp (hunk 4 above)
Comment (est77) [reply sandbox/samples/cpp/sphereflake/sphereflake.cpp]: We could, but the original code from http://www.realtimerendering.com/resources/SPD/ is pretty old and I thought it was too much work to modernize it just for a sample.
I was surprised it even compiles now (with some minor fixes).

**Pull request merged**